File: LeetCode/200-number-of-islands/ex_200.java
public class Solution {
    public int numIslands(char[][] grid) {
        int count = 0;
        for (int x = 0; x < grid.length; x++) {
            for (int y = 0; y < grid[0].length; y++) {
                if (grid[x][y] == '1') {
                    count++;
                    dfs(grid, x, y);
                }
            }
        }
        return count;
    }
    
    private void dfs(char[][] grid, int x, int y) {
        if (x < 0 || y < 0 || x >= grid.length || y >= grid[0].length || grid[x][y] != '1')    return;
        grid[x][y] = '0';
        for (int i = 0; i < 4; i++) {
            dfs(grid, x + direction[i][0], y + direction[i][1]);
        }
    }
    
    private static final int[][] direction = new int[][] {{0, 1}, {1, 0}, {-1, 0}, {0, -1}};
}

File: LeetCode/200-number-of-islands/ex_200.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        def dfs(x, y):
            if x < 0 or y < 0 or x >= len(grid) or y >= len(grid[0]) or grid[x][y] != '1':   return
            grid[x][y] = '0'
            for i in range(4):
                dfs(x + shift[i][0], y + shift[i][1])
        
        if not grid or not grid[0]: return 0
        count = 0
        shift = [[1,0],[0,1],[-1,0],[0,-1]]
        for x in range(len(grid)):
            for y in range(len(grid[0])):
                if grid[x][y] == '1':
                    dfs(x,y)
                    count += 1
        return count